quinn-proto/src/connection/mtud.rs

```diff
@@ -131,31 +128,31 @@ impl MtuDiscovery {
     }
 
     /// Notifies the [`MtuDiscovery`] that a non-probe packet was lost
-    pub(crate) fn on_non_probe_lost(&mut self, now: Instant, packet_number: u64, packet_size: u16) {
-        // Ignore packets smaller or equal than the base size
-        if packet_size <= BASE_UDP_PAYLOAD_SIZE {
-            return;
-        }
-
-        // Ignore lost packets if we have received an ACK for a more recent MTU-sized packet
-        if self
-            .largest_acked_mtu_sized_packet
-            .map_or(false, |pn| packet_number < pn)
-        {
-            return;
-        }
+    ///
+    /// When done notifying of lost packets, [`MtuDiscovery::detect_black_hole`] must be called, to
+    /// ensure the last loss burst is properly processed and to trigger black hole recovery logic if
+    /// necessary
+    pub(crate) fn on_non_probe_lost(&mut self, packet_number: u64, packet_bytes: u16) {
+        self.black_hole_detector
+            .on_non_probe_lost(packet_number, packet_bytes);
+    }
 
-        // The packet counts towards black hole detection
-        self.black_hole_counter += 1;
-        if self.black_hole_counter > BLACK_HOLE_THRESHOLD {
-            self.black_hole_counter = 0;
-            self.largest_acked_mtu_sized_packet = None;
+    /// Returns true if a black hole was detected
+    ///
+    /// Calling this function will close the previous loss burst. If a black hole is detected, the
+    /// current MTU will be reset to `BASE_UDP_PAYLOAD_SIZE`.
+    pub(crate) fn detect_black_hole(&mut self, now: Instant) -> bool {
+        if self.black_hole_detector.detect_black_hole() {
```

Comment:
Style nit: consider inverting the check here so you can early return for the other case. I'd probably aslo change the method name (as well as the one one `BlackHoleDetector`) to something like `black_hole_detected()`, since it doesn't seem like this method is actually running much logic to "detect" a black hole.